dash/include/dash/Graph.h

```diff
@@ -0,0 +1,876 @@
+    _glob_mem_edge->add_globmem(*_glob_mem_out_edge);
+    _glob_mem_edge->add_globmem(*_glob_mem_in_edge);
+    // Register deallocator at the respective team instance
+    _team->register_deallocator(this, std::bind(&Graph::deallocate, this));
+
+    _vertices = vertex_it_wrapper(this);
+    _in_edges = in_edge_it_wrapper(this);
+    _out_edges = out_edge_it_wrapper(this);
+    _edges = edge_it_wrapper(this);
+    return true;
+  }
+
+  /**
+   * Deallocates global memory.
+   */
+  void deallocate() {
+    if(_glob_mem_vertex != nullptr) {
+      delete _glob_mem_vertex;
+      _glob_mem_vertex = nullptr;
+    }
+    if(_glob_mem_in_edge != nullptr) {
+      delete _glob_mem_in_edge;
+      if(_glob_mem_in_edge == _glob_mem_out_edge) {
+        _glob_mem_out_edge = nullptr;
+      }
+      _glob_mem_in_edge = nullptr;
+    }
+    if(_glob_mem_out_edge != nullptr 
+        && _glob_mem_out_edge != _glob_mem_in_edge) {
+      delete _glob_mem_out_edge;
+      _glob_mem_out_edge = nullptr;
+    }
+    if(_glob_mem_edge != nullptr) {
+      delete _glob_mem_edge;
+      _glob_mem_edge = nullptr;
+    }
+    if(_team != nullptr) {
+      // Remove deallocator from the respective team instance
+      _team->unregister_deallocator(this, std::bind(&Graph::deallocate, this));
+    }
+  }
+
+  /**
+   * Returns the team containing all units associated with this container
+   */
+  Team & team() const {
+    return *_team;
+  }
+
+private:
+
+  /**
+   * Inserts an edge locally. The source vertex must belong to this unit.
+   * 
+   * \return Local iterator to the created edge.
+   */
+  template<typename TargetIterator>
+  typename glob_mem_edge_type::local_iterator add_local_edge(      
+      const local_vertex_iterator source, 
+      const TargetIterator target, 
+      const EdgeProperties & prop,
+      glob_mem_edge_type * glob_mem_edge
+  ) {
+    auto edge = edge_type(source, target, prop, _myid);
+    return glob_mem_edge->push_back(source.pos(), edge);
+  }
+
+  /**
+   * Inserts an edge locally. The source vertex must belong to this unit.
+   * 
+   * \return Local iterator to the created edge.
+   */
+  typename glob_mem_edge_type::local_iterator add_local_edge(      
+      const vertex_index_type source, 
+      const vertex_index_type target, 
+      const EdgeProperties & prop,
+      glob_mem_edge_type * glob_mem_edge
+  ) {
+    auto edge = edge_type(source, target, prop);
+    return glob_mem_edge->push_back(source.offset, edge);
+  }
+
+  /*
+  team_unit_t vertex_owner(vertex_size_type v, vertex_size_type n_vertices) {
+    int owner_id = static_cast<double>(v) / (static_cast<double>(n_vertices) / 
+        _team->size());
+    int size = _team->size();
+    int owner_id = static_cast<double>(v) / (n_vertices
+         * ((static_cast<double>(_myid) + 1) / (size * (size + 1) / 2)));
+    team_unit_t owner { owner_id };
+    return owner;
+  }
+  */
+
+  // TODO: Generalize following methods for other edge attributes than { int }
```

Comment:
does this limit the usability of the whole Graph, or are these just convenience functions for the user? I.e, they are not used by the implementation itself?